### ISSUE TITLE
BC policy

### ISSUE THREAD

**Issue by yguedidi** [facebook/facebook-php-sdk-v4#436]
I just want to discuss about 5.0.

As we do not tag any 5.0+ release, do we really need to keep BC with 4.1?
I think it's the responsability of developers to **never** use non-stable packages.

I know that we have 4.1 for a while, so keeping a dedicated branch to users who require that version make sense, but what about 5.0?
We just create it, I don't think there are devs who use it.

@gfosco @SammyK thoughts?

**Comment by SammyK** [comment]
> I think it's the responsability of developers to never use non-stable packages.

That's probably true for production environments, but at the very least devs should accept that development branches can BC on the daily.

I've been telling people that v4.1/v5 has had a feature freeze... but there was nothing official marking it as such. I just thought the docs were going to get posted soon, so that was just my bad for announcing it too soon. I just got too excited! :D

> I don't think there are devs who use it.

I know of a few who do and I think there are some people who are watching this repo who use it and just fix their code when there's a BC. Although I could be wrong.

Any of you silent watchers use the v4.1/v5 dev branch? :)

**Comment by yguedidi** [comment]
@SammyK Oups sorry, I forgot to mention that yes, a talk about production.

> at the very least devs should accept that development branches can BC on the daily.

Then we still can break things before a stable release :)

Depends on @gfosco decision, I will open many issues to discuss my thoughts about some refactoring I think will be nice to have for 5.0

**Comment by SammyK** [comment]
:+1:

**Comment by ajessu** [comment]
> I know of a few who do and I think there are some people who are watching this repo who use it and just fix their code when there's a BC. Although I could be wrong.
> 
> Any of you silent watchers use the v4.1/v5 dev branch? :)

I'm one of those (the `4.1-dev` branch, which is unreleased/untagged also). I'm using the branch knowing that it's a non-released branch, and that BC are expected, so I'm following changes in the repo closely. I wouldn't expect a non-released branch to be stable, or follow a BC policy, so agree with everything here.

I'd think ongoing non-BC changes would be expected on the 4.1-dev branch or on a newer 5.0-dev branch, as they would essentially be the same. The break is from `4.0.x` to `5.0`, not from `4.1` to `5.0`, as `4.1` was never released, it's only been an alias for master development.

On our case, we're ensuring not to hit non-bc code on the dev branch, by pointing at the repo like so:
`"facebook/php-sdk-v4" : "dev-master#<hash of commit that has been tested>"`, which locks that even in the case of running `composer update` on the project, and have been moving hash reference up in history, only after testing, and would be doing so until a stable release is reached.

In our case, I'd expect adjusting code for any future non-BC change on the branch is an expected inconvenience, but IMHO that's the price to pay to be able to use the much better version of the library before it's released :+1:.

That said, I'd expect a sooner than later release, as the branch seems like it's ready, though still curious about @yguedid's refactoring ideas.

**Comment by SammyK** [comment]
Thanks for the excellent feedback @ajessu! I think the `"dev-master#<hash of commit that has been tested>"` idea is really excellent advice for those using a dev branch.

> though still curious about @yguedidi 's refactoring ideas.

Me too! :)

**Comment by yguedidi** [comment]
@SammyK @ajessu have a look to #437 :)

**Comment by gfosco** [comment]
Sorry I've not been very present on this repo lately.  I'm open to jamming more stuff in 5.0 if you all want to, or we can try to get 5.0 out the door.  I still haven't found time to load the docs.

**Comment by SammyK** [comment]
>  or we can try to get 5.0 out the door.

:shipit: Let's ship it! :)

**Comment by yguedidi** [comment]
We should keep the door closed (#447) :)

**Comment by gfosco** [comment]
Closing this.. if there are still points to discuss, please start a new issue.
